virtualizarr/manifests/registry.py

```diff
@@ -0,0 +1,236 @@
+"""
+Based on https://docs.rs/object_store/0.12.2/src/object_store/registry.rs.html#176-218
+
+This module may eventually be upstreamed into https://github.com/developmentseed/obstore
+"""
+
+from __future__ import annotations
+
+from collections import namedtuple
+from typing import TYPE_CHECKING, Dict, Iterator, Optional, Tuple, TypeAlias
+from urllib.parse import urlparse
+
+if TYPE_CHECKING:
+    from obstore.store import (
+        ObjectStore,
+    )
+
+Url: TypeAlias = str
+Path: TypeAlias = str
+
+UrlKey = namedtuple("UrlKey", ["scheme", "netloc"])
+
+
+def get_url_key(url: Url) -> UrlKey:
+    """
+    Generate the UrlKey containing a url's scheme and authority/netloc that is used a the
+    primary key's in a [ObjectStoreRegsitry.map][virtualizarr.manifests.ObjectStoreRegistry.map]
+
+    Parameters
+    ----------
+    url
+        Url to generate a UrlKey from
+
+    Returns
+    -------
+        NamedTuple containing the Url's scheme and authority/netloc
+
+    Raises
+    ------
+    ValueError
+        If provided Url does not contain a scheme based on [urllib.parse.urlparse][]
+    """
+    parsed = urlparse(url)
+    if not parsed.scheme:
+        raise ValueError(
+            f"Urls are expected to contain a scheme, received {url} which parsed to {parsed}"
+        )
+    return UrlKey(parsed.scheme, parsed.netloc)
+
+
+class PathEntry:
```

Comment:
I'm not totally sure I understand why this recursive(?) data structure is necessary. Couldn't we achieve similar behaviour just by having a flat dict of `str` paths and first matching on only the first segment, then if that fails matching on `/segment1/segment2` etc.? i.e. using `.startswith()` or something like that?

Reply:
I ported this module from Rust - https://docs.rs/object_store/latest/src/object_store/registry.rs.html#137 and went with their design. I expect their primary motivations for the nested structure is `O(path_depth)` vs `O(total_registrations)` lookup performance and de-duplication of common prefixes.